File: Algorithms/Swift/Easy/242_Valid_Anagram.swift
// Hash Table, Time Complexity O(n), n is the length of the longer string
class Solution {
    func isAnagram(s: String, _ t: String) -> Bool {
        if s.characters.count != t.characters.count {
            return false
        }

        var dict = [Character : Int]()
        for character in s.characters {
            if dict[character] != nil {
                dict[character]! += 1
            } else {
                dict[character] = 1
            }
        }

        for character in t.characters {
            if dict[character] == nil {
                return false
            } else {
                dict[character]! -= 1
            }
        }

        for counter in dict.values {
            if counter != 0 {
                return false
            }
        }

        return true
    }
}

// Sort, Time Complexity O(nlogn), n is the length of the longer string
class Solution {
    func isAnagram(s: String, _ t: String) -> Bool {
        if s.characters.count != t.characters.count {
            return false
        }

        let sortedS = String(s.characters.sort() { $0 < $1 })
        let sortedT = String(t.characters.sort() { $0 < $1 })

        return sortedS == sortedT
    }
}

File: Algorithms/JavaScript/Easy/242_Valid_Anagram.js
// Hash Table
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length === 0) {
        if (t.length === 0) {
            return true
        } else {
            return false
        }
    }

    if (s.length != t.length) {
        return false
    }

    var hash = {}
    for (i = 0; i < s.length; i++) {
        let character = s[i]
        if (!hash[character]) {
            hash[character] = 1
        } else {
            hash[character] += 1
        }
    }

    for (i = 0; i < t.length; i++) {
        let character = t[i]
        if (!hash[character]) {
            return false
        } else if (hash[character] === 0) {
            return false
        }
        hash[character] -= 1
    }
    return true
};

// Sort
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length === 0) {
        if (t.length === 0) {
            return true
        } else {
            return false
        }
    }

    if (s.length != t.length) {
        return false
    }

    s = s.split("").sort(strCompare).join("")
    t = t.split("").sort(strCompare).join("")

    return s === t
};

function strCompare(str1, str2) {
    return str1.localeCompare(str2)
}
